collect_app/src/main/java/org/odk/collect/android/widgets/items/SelectOneFromMapDialogFragment.kt

```diff
@@ -103,43 +131,60 @@ internal class SelectChoicesMapData(
     scheduler: Scheduler,
     prompt: FormEntryPrompt,
     private val selectedIndex: Int?
-) :
-    SelectionMapData {
+) : SelectionMapData {
 
     private val mapTitle = MutableLiveData(prompt.longText)
-    private val itemCount = MutableLiveData<Int>()
+    private val itemCount = MutableNonNullLiveData(0)
     private val items = MutableNonNullLiveData(emptyList<MappableSelectItem>())
     private val isLoading = MutableNonNullLiveData(true)
+    private val invalidGeometry = MutableLiveData<InvalidGeometry?>(null)
 
     init {
         isLoading.value = true
 
         scheduler.immediate(
             background = {
-                val selectChoices = prompt.selectChoices
-                val itemCount = selectChoices.size
-                val items: List<MappableSelectItem> = loadItemsFromChoices(selectChoices, prompt)
-
-                Pair(itemCount, items)
+                loadItemsFromChoices(prompt.selectChoices, prompt)
             },
             foreground = {
-                itemCount.value = it.first
-                items.value = it.second
-                isLoading.value = false
+                when (it) {
+                    is Result.Success -> {
+                        itemCount.value = prompt.selectChoices.size
+                        items.value = it.value
+                        isLoading.value = false
+                    }
+
+                    is Result.Error -> {
+                        invalidGeometry.value = it.value
+                    }
+                }
             }
         )
     }
 
     private fun loadItemsFromChoices(
         selectChoices: MutableList<SelectChoice>,
         prompt: FormEntryPrompt
-    ): List<MappableSelectItem> {
-        return selectChoices.foldIndexed(emptyList()) { index, list, selectChoice ->
+    ): Result<List<MappableSelectItem>, InvalidGeometry> {
+        return selectChoices.foldIndexed(emptyList<MappableSelectItem>()) { index, list, selectChoice ->
             val geometry = selectChoice.getChild("geometry")
 
             if (geometry != null) {
-                val latitude = geometry.split(" ")[0].toDouble()
-                val longitude = geometry.split(" ")[1].toDouble()
+                val latitude: Double
+                val longitude: Double
+
+                try {
+                    latitude = geometry.split(" ")[0].toDouble()
+                    longitude = geometry.split(" ")[1].toDouble()
+
+                    if (latitude.absoluteValue > 90 || longitude.absoluteValue > 180) {
+                        return InvalidGeometry(prompt.getSelectChoiceText(selectChoice), geometry)
```

Comment:
I'm really not happy with this (and the below) early return but haven't really come up with anything better 🤔